Add support for configuring static sampling strategies via file

chancez opened the pull request:
Closes #491

Adds support for https://www.jaegertracing.io/docs/1.17/sampling/#collector-sampling-configuration

I'm opening this now while I test this with an actual application reading the file from disk, but I wanted to open it for feedback, or in case someone else wants to steward the change in the event that I get busy!

joe-elliott commented on config/config.go:
```diff
@@ -392,10 +399,20 @@ func (sc *SamplerConfig) NewSampler(
+			options = append(
+				options,
+				jaeger.SamplerOptions.SamplingServerURL(sc.SamplingServerURL),
+				jaeger.SamplerOptions.SamplingRefreshInterval(sc.SamplingRefreshInterval),
```
We should keep the refresh interval on the file option.  It will allow people to dynamically swap out the sampling file w/o restarting their service.

chancez replied:
Oh I originally intended that, didn't catch this, thanks!

joe-elliott commented on sampler_remote.go:
```diff
@@ -297,6 +297,14 @@ func (u *AdaptiveSamplerUpdater) Update(sampler SamplerV2, strategy interface{})
 // -----------------------
 
+type fileSamplingStrategyFetcher struct {
+	strategilesFile string
```
nit: strategiesFile

yurishkuro commented on config/config.go:
```diff
@@ -392,11 +399,19 @@ func (sc *SamplerConfig) NewSampler(
 			jaeger.SamplerOptions.OperationNameLateBinding(sc.OperationNameLateBinding),
 			jaeger.SamplerOptions.SamplingRefreshInterval(sc.SamplingRefreshInterval),
 		}
+		if sc.SamplerStrategiesFile != "" {
```
nit: this logic establishes implicit priority of fields. Instead, I would suggest returning an error if both File and URL are not blanks.

yurishkuro commented on config/config.go:
```diff
@@ -89,6 +89,13 @@ type SamplerConfig struct {
 
+	// SamplerStrategiesFile is this services static sampling strategy
+	// configuration.
+	// Conflicts with SamplingServerURL and SamplingRefreshInterval.
```
SamplingRefreshInterval is still relevant

yurishkuro reviewed (commented):
please add unit tests for file fetcher and new configuration logic

chancez commented:
Working through the review items. Also missed support for the env-vars config. Should it be `JAEGER_SAMPLER_STRATEGIES_FILE` or `JAEGER_SAMPLING_STRATEGIES_FILE` (same for the struct fields)?

yurishkuro commented:
`JAEGER_SAMPLER_*`, to be consistent with other vars.

chancez commented:
@yurishkuro do you have a recommendation on testing the file fetcher? The HTTP tester isn't actually used in the tests, it's mocked out with the `testSamplingStrategyFetcher`.

chancez commented:
Pushed some fixups (needs to be rebased). I added tests for config validation as well and added env-var handling. I'm still looking at how to best write a test for the fileFetcher.

chancez commented:
Added a very test for the fileFetcher. Let me know if things look good and I can squash.

yurishkuro commented on config/config_test.go:
```diff
@@ -182,6 +182,37 @@ func TestSamplerConfig(t *testing.T) {
+	setEnv(t, envSamplingEndpoint, "http://themaster:5778/sampling")
+	setEnv(t, envSamplerMaxOperations, "10")
+	setEnv(t, envSamplerRefreshInterval, "1m1s")                // 61 seconds
+	setEnv(t, envSamplerStrategiesFile, "/tmp/strategies.json") // 61 seconds
```
```suggestion
	setEnv(t, envSamplerStrategiesFile, "/tmp/strategies.json")
```

yurishkuro reviewed (commented):
This test seems unrelated, but it does not fail in the main branch.
```
=== RUN   TestHTTPTransport
    http_test.go:60: never received a span
--- FAIL: TestHTTPTransport (2.00s)
```

chancez commented:
Looks like a flake?

chancez commented:
Squashed & Rebased with all requested changes.

yurishkuro commented:
There's noticeable code coverage drop from this change, can we add more tests?

chancez commented:
@yurishkuro Ok, I got the coverage back up, and everything is passing.

yurishkuro commented on config/config_test.go:
```diff
@@ -569,6 +656,35 @@ func TestInvalidSamplerType(t *testing.T) {
 	rcs.Close()
 }
 
+func TestInvalidRemoteSamplerStrategySettings(t *testing.T) {
```
isn't this already tested in TestInvalidSamplerConfig?

chancez replied:
No that doesn't use `cfg.NewSampler`, it uses `cfg.samplerConfigFromEnv` which are different code paths with different errors.

yurishkuro commented on sampler_remote_test.go:
```diff
@@ -599,3 +601,50 @@ func TestRemotelyControlledSampler_printErrorForBrokenUpstream(t *testing.T) {
 }
+
+func TestRemotelyControlledSampler_samplingStrategiesFile(t *testing.T) {
+	content := []byte("probabilistic")
```
I am confused by this - a single word "probabilistic" is not a valid sampling strategy, how does this work?

chancez replied:
This was based on how other tests were doing it ¯\_(ツ)_/¯. I can take another look. Like I said previously, I can't find any tests which actually stressed the existing remote sampling logic.

chancez replied:
I wasn't getting coverage for `SamplerOptions.SamplingStrategiesFile` without it, because the other tests use the remote HTTP fetcher (eg: TestRemoteSamplerOptions).

yurishkuro replied:
Can that part be tested in isolation? If you want to test it with whole Remote sampler, then the policy in the file needs to be (a) parseable, and (b) different from the default sampler, so that the test can assert end-to-end policy application.

chancez replied:
I'm working on using a full valid config. I actually am noticing that there may need to be client side logic to find the right section of the config for the provided serviceName. I'm looking at the upstream jaeger project to figure out how the collector/agent does this.

yurishkuro replied:
There is a difference between config that is retrieved by SDK from the backend and the config that the backend has. The former is always scoped to the service making the request for sampling strategies, so the response would not contain a map for different services. Specifically, the API between remote sampler and the backend is the JSON version of this: https://github.com/jaegertracing/jaeger-idl/blob/34396033ff11c60fced342ab2858ace278fedaa8/thrift/sampling.thrift#L50

yurishkuro commented on sampler_remote_test.go:
```diff
@@ -599,3 +601,50 @@ func TestRemotelyControlledSampler_printErrorForBrokenUpstream(t *testing.T) {
+
+	assert.Equal(t, sampler.samplerOptions.samplingStrategiesFile, filename)
+
+	_, ok := sampler.Sampler().(*ProbabilisticSampler)
```
I think this does not prove that the sampler was actually instantiated per the config file, because Remote defaults to probabilistic anyway. I think there is a way to pass a different init sampler to Remote one, or you can vary which strategy is provided in the config file.

yurishkuro reviewed (commented):
Please also add a section to README explaining how to use this new feature with an example of config file.

chancez commented:
I guess the way upstream does this, is using https://github.com/jaegertracing/jaeger/tree/master/plugin/sampling/strategystore/static which is also what the opentelemetry jaeger-receiver uses. Perhaps I should refactor around using this?

yurishkuro commented:
closing as stale
